components/script/dom/bindings/root.rs

```diff
@@ -589,6 +590,22 @@ impl<T: DomObject + PartialEq> PartialEq<T> for MutDom<T> {
     }
 }
 
+#[cfg(debug_assertions)]
+pub(crate) fn assert_in_script() {
+    debug_assert!(thread_state::get().is_script());
+}
+
+#[cfg(debug_assertions)]
+pub(crate) fn assert_in_layout() {
+    debug_assert!(thread_state::get().is_layout());
+}
+
+#[cfg(not(debug_assertions))]
+pub(crate) fn assert_in_layout() {}
+
+#[cfg(not(debug_assertions))]
+fn assert_in_script() {}
```

Comment:
Those `#[cfg]` attributes and separate function definitions look redundant. The expansion of `debug_assert!` (instead of `assert!`) is `if false { … }` when `debug_assertions` is not enabled, which is easy for the optimizer to eliminate.